src/wp-includes/block-template-utils.php

```diff
@@ -1326,19 +1326,60 @@ function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '
 	}
 
 	$template_hierarchy = array( $slug );
-
 	// Most default templates don't have `$template_prefix` assigned.
-	if ( $template_prefix ) {
-		list( $type ) = explode( '-', $template_prefix );
-		// These checks are needed because the `$slug` above is always added.
+	if ( ! empty( $template_prefix ) ) {
+		list($type) = explode( '-', $template_prefix );
+		// We need these checks because we always add the `$slug` above.
 		if ( ! in_array( $template_prefix, array( $slug, $type ), true ) ) {
 			$template_hierarchy[] = $template_prefix;
 		}
 		if ( $slug !== $type ) {
 			$template_hierarchy[] = $type;
 		}
+	} else {
+		$matches = array();
+		if ( preg_match( '/^(author|category|archive|tag|page)-(.+)$/', $slug, $matches ) ) {
+			$template_hierarchy[] = $matches[1];
+		} elseif ( preg_match( '/^(single|taxonomy)-(.+)$/', $slug, $matches ) ) {
+			$type           = $matches[1];
+			$slug_remaining = $matches[2];
+			if ( 'single' === $type ) {
+				$post_types = get_post_types();
+				foreach ( $post_types as $post_type ) {
+					if ( str_starts_with( $slug_remaining, $post_type ) ) {
+						// If $slug_remaining is equal to $post_type we have the single-$post_type template.
+						if ( $slug_remaining === $post_type ) {
+							$template_hierarchy[] = 'single';
+							break;
+						}
+						// If $slug_remaining is single-$post_type-$slug template.
+						if ( str_starts_with( $slug_remaining, $post_type . '-' ) && strlen( $slug_remaining ) > strlen( $post_type ) + 1 ) {
+							$template_hierarchy[] = "single-$post_type";
+							$template_hierarchy[] = 'single';
+							break;
+						}
+					}
+				}
+			} elseif ( 'taxonomy' === $type ) {
+				$taxonomies = get_taxonomies();
+				foreach ( $taxonomies as $taxonomy ) {
+					if ( str_starts_with( $slug_remaining, $taxonomy ) ) {
+						// If $slug_remaining is equal to $taxonomy we have the taxonomy-$taxonomy template.
+						if ( $slug_remaining === $taxonomy ) {
+							$template_hierarchy[] = 'taxonomy';
+							break;
+						}
+						// If $slug_remaining is taxonomy-$taxonomy-$term template.
+						if ( str_starts_with( $slug_remaining, $taxonomy . '-' ) && strlen( $slug_remaining ) > strlen( $taxonomy ) + 1 ) {
+							$template_hierarchy[] = "taxonomy-$taxonomy";
+							$template_hierarchy[] = 'taxonomy';
+							break;
+						}
+					}
+				}
+			}
+		}
```

Comment:
I'm wondering if this whole block can be made a bit more readable and efficient.  There's a good bit of nesting here, and it could be improved with some early bails.  E.g.

```suggestion
		preg_match( '/^(author|category|archive|tag|page|taxonomy|single)-(.+)$/', $slug, $matches );
		
		if ( ! $matches ) {
		    return $template_hierarchy;
		}
		
		$type = $matches[1];
		
		if ( in_array( $type, array( 'author', 'category', 'archive', 'tag', 'page' ), true ) ) {
		    $template_hierarchy[] = $type;		    
		    
		    return $template_hierarchy;
		}		
		
		$slug_remaining = $matches[2];		
		
		if ( 'single' === $type ) {
			$post_types = get_post_types();			
			
			foreach ( $post_types as $post_type ) {
				if ( ! str_starts_with( $slug_remaining, $post_type ) ) {
				        continue;
				} 
				
				// If $slug_remaining is equal to $post_type we have the single-$post_type template.
				if ( $slug_remaining === $post_type ) {
					$template_hierarchy[] = 'single';
					break;
				}
				
				
				// If $slug_remaining is single-$post_type-$slug template.
				if ( strlen( $slug_remaining ) > strlen( $post_type ) + 1 ) {
					$template_hierarchy[] = "single-$post_type";
					$template_hierarchy[] = 'single';
					break;
				}				
			}
		} elseif ( 'taxonomy' === $type ) {
			$taxonomies = get_taxonomies();			
			
			foreach ( $taxonomies as $taxonomy ) {
				if ( ! str_starts_with( $slug_remaining, $taxonomy ) ) {
				        continue;				        
				}
				
				// If $slug_remaining is equal to $taxonomy we have the taxonomy-$taxonomy template.
				if ( $slug_remaining === $taxonomy ) {
					$template_hierarchy[] = 'taxonomy';
					break;
				}				
				
				// If $slug_remaining is taxonomy-$taxonomy-$term template.
				if ( strlen( $slug_remaining ) > strlen( $taxonomy ) + 1 ) {
					$template_hierarchy[] = "taxonomy-$taxonomy";
					$template_hierarchy[] = 'taxonomy';
					break;
				}
			}
		}
```

I apologize if the whitespace is off here(doing it on GH).  This combines the regex into one liner and bails early.  Similar improvements throughout the block.  The code below this suggestion could be moved above this so that $template_hierarchy` can be an early return.  If these are expected to returned in a specific order, the code would be different.

I also wonder if it could be condensed even further by combining the `get_post_types()` and `get_taxonomies()` `foreach` blocks into one loop with a combined array of both types.  Then, the `$template_hierarchy[]` could append the `$type` for each.

Reply:
Hi @dream-encode, thank you for the feedback. I applied the ideas you suggested. We could not return early because we must unconditionally execute the logic after all the if-else clauses. But I followed some simplifications suggested and condense the logic for post types and taxonomies.

Reply:
@jorgefilipecosta I understand.  I still think there's room for improvement here.  Note the line numbers I am referencing here are after applying this latest patch to `trunk`.

Is the `$template_hierarchy` array returned from this function expected to be in a specific order?  I could only find one call in Core, and that seems to just be getting this data for a REST request made by the editor.  So, I am not sure if the Editor is expecting these to be in a specific, descending order or not.

If not, we could move lines 1374-1396( `// Handle `archive` template.` to the last `$template_hierarchy[] = 'index';` above this block but right below `$template_hierarchy = array( $slug );`.  

We could then make `if ( ! empty( $template_prefix ) ) {` its own block and return early from there.

Then, that un-nests the `preg_match( '/^(author|category|archive|tag|page|taxonomy|single)-(.+)$/', $slug, $matches );` and again can be early-returned if `! $matches`.